AWSAuthSDK/Sources/AWSFacebookSignIn/AWSFacebookSignInProvider.m

```diff
@@ -111,30 +89,38 @@ - (NSString *)identityProviderName {
 - (AWSTask<NSString *> *)token {
     Class fbSDKAccessToken = NSClassFromString(@"FBSDKAccessToken");
     NSString *tokenString = [fbSDKAccessToken currentAccessToken].tokenString;
-    NSDate *idTokenExpirationDate = [fbSDKAccessToken currentAccessToken].expirationDate;
+    BOOL isTokenExpired = [fbSDKAccessToken currentAccessToken].dataAccessExpired;
     
     if (tokenString
         // If the cached token expires within 10 min, tries refreshing a token.
-        && [idTokenExpirationDate compare:[NSDate dateWithTimeIntervalSinceNow:AWSFacebookSignInProviderTokenRefreshBuffer]] == NSOrderedDescending) {
+        && !isTokenExpired) {
         return [AWSTask taskWithResult:tokenString];
     }
     
-    _taskCompletionSource = [AWSTaskCompletionSource taskCompletionSource];
-    Class fbSDKLoginManager = NSClassFromString(@"FBSDKLoginManager");
-    [fbSDKLoginManager renewSystemCredentials:^(ACAccountCredentialRenewResult result, NSError *error) {
-        if (result == ACAccountCredentialRenewResultRenewed) {
-            
+    self.taskCompletionSource = [AWSTaskCompletionSource taskCompletionSource];
+
+    UIViewController *vc = [[UIViewController alloc] init];
+    if (!self.facebookLogin){
+        [self createFBSDKLoginManager];
+    }
+    
+    [self.facebookLogin reauthorizeDataAccess: vc
+                                      handler:^(FBSDKLoginManagerLoginResult *result, NSError *error) {
+        
+        if (error) {
+            self.taskCompletionSource.error = error;
+        } else {
             Class fbSDKAccessToken = NSClassFromString(@"FBSDKAccessToken");
             if (fbSDKAccessToken) {
-                _taskCompletionSource.result = [fbSDKAccessToken currentAccessToken].tokenString;
+                self.taskCompletionSource.result = [fbSDKAccessToken currentAccessToken].tokenString;
             } else {
                 AWSDDLogError(@"Error setting facebook token to result");
```

Comment:
not sure what you are referring to here, it is an NSError and is being set up above in the first part of the if statement. are you talking about the else after if(fbSDKAcessToken) ? I think maybe so

Reply:
In Line 117:

```
AWSDDLogError(@"Error setting facebook token to result");
```

Change this from a log statement to creating an NSError (using similar patterns as elsewhere, probably an "unknown" error type), and then setting `self.taskCompletionSource.error` to be the new error. That way, instead of simply logging an error, we actually will trigger the callback with an error condition